cli/paas/install_client.go

```diff
@@ -21,88 +38,92 @@ type InstallClient struct {
 }
 
 // Install deploys carrier to the cluster.
-func (c *InstallClient) Install(cmd *cobra.Command, deployments *kubernetes.DeploymentSet) error {
+func (c *InstallClient) Install(cmd *cobra.Command, options *kubernetes.InstallationOptions) error {
 	c.ui.Note().Msg("Carrier installing...")
 
-	// Hack? Override static default for system domain with a
-	// function which queries the cluster for the necessary
-	// data. If that data could not be found the system will fall
-	// back to cli option and/or interactive entry by the user.
-	//
-	// NOTE: This is function is set here and not in the gitea
-	// definition because the function has to have access to the
-	// cluster in question, and that information is only available
-	// now, not at deployment declaration time.
-
-	domain, err := deployments.NeededOptions.GetOpt("system_domain", "")
+	var err error
+	options, err = options.Populate(kubernetes.NewCLIOptionsReader(cmd))
 	if err != nil {
 		return errors.Wrap(err, "Couldn't install carrier")
 	}
-	domain.DynDefaultFunc = func(o *kubernetes.InstallationOption) error {
-		ips := c.kubeClient.GetPlatform().ExternalIPs()
-		if len(ips) > 0 {
-			domain := fmt.Sprintf("%s.nip.io", ips[0])
-			o.Value = domain
-		}
-		// else leave invalid, to be handled by cli option
-		// reader or interactive entry
-		return nil
-	}
-
-	deployments.PopulateNeededOptions(kubernetes.NewCLIOptionsReader(cmd))
 
 	interactive, err := cmd.Flags().GetBool("interactive")
 	if err != nil {
 		return errors.Wrap(err, "Couldn't install carrier")
 	}
 
 	if interactive {
-		deployments.PopulateNeededOptions(kubernetes.NewInteractiveOptionsReader(os.Stdout, os.Stdin))
+		options, err = options.Populate(kubernetes.NewInteractiveOptionsReader(os.Stdout, os.Stdin))
+		if err != nil {
+			return errors.Wrap(err, "Couldn't install carrier")
+		}
 	} else {
-		deployments.PopulateNeededOptions(kubernetes.NewDefaultOptionsReader())
+		options, err = options.Populate(kubernetes.NewDefaultOptionsReader())
+		if err != nil {
+			return errors.Wrap(err, "Couldn't install carrier")
+		}
 	}
 
-	c.showInstallConfiguration(deployments)
+	c.showInstallConfiguration(options)
 
 	// TODO (post MVP): Run a validation phase which perform
 	// additional checks on the values. For example range limits,
 	// proper syntax of the string, etc. do it as pghase, and late
 	// to report all problems at once, instead of early and
 	// piecemal.
 
-	err = c.deploySet(deployments)
+	deployment := deploymentSet["traefik"]
+	err = deployment.Deploy(c.kubeClient, c.ui, options.ForDeployment(deployment.ID()))
```

Comment:
IMHO the whole `options.ForDeployment(deployment.ID())` can be moved into the `Deploy` function.
Plus removing the temp variable the section could be reduced to:
```
err = deploymentSet["traefik"].Deploy(c.kubeClient, c.ui, options)
```
This should simplify the coming loop over the remainder also.

Reply:
true. The idea was that the caller would give each Deployment only the relevant options (so they wouldn't even have access to things they shouldn't). Imagine someone needing an option in a Deployment, and simply because the whole options object is available, they would bypass the `ForDeployment` method and read the option of another deployment directly.
In that case, we wouldn't be able to tell which option is needed by which deployment simply by looking at the options object.

Of course, with only one option currently being used maybe this measure is meaningless, but it could be useful in the future. What do you think?

Reply:
Ok. I am not to fussed actually. It simply looked not as clean as I thought it could be. Leaving the argument as-is is fine. and the argument makes sense, so plus one on that too.

I would still go with slices and direct iteration over the map.
If we keep the map, I would at least drop the temp variable.